cayleypy/algo/beam_search.py

```diff
@@ -121,3 +184,155 @@ def _restore_path(found_layer_id: int) -> Optional[list[int]]:
 
         # Path not found.
         return BeamSearchResult(False, 0, None, debug_scores, graph.definition)
+
+    def search_advanced(
+        self,
+        start_state: AnyStateType,
+        destination_state: Optional[AnyStateType] = None,
+        *,
+        beam_width: int = 1000,
+        max_steps: int = 1000,
+        history_depth: int = 0,
+        predictor: Optional[Predictor] = None,
+        verbose: int = 0,
+    ) -> BeamSearchResult:
+        """Advanced beam search using PyTorch with non-backtracking capabilities.
+
+        This method implements an improved beam search algorithm that supports:
+        - Non-backtracking constraints (avoiding revisiting states)
+        - Various prediction models (torch models, sklearn models, heuristics)
```

Comment:
This is not something new in this algorithm compared to "simple". This variability of models is the encapsulated in the Predictor class.